unumpy/linalg/_multimethods.py

```diff
@@ -0,0 +1,203 @@
+import functools
+import operator
+from uarray import create_multimethod, mark_as, all_of_type, Dispatchable
+import builtins
+
+create_numpy = functools.partial(create_multimethod, domain="numpy.linalg")
+
+from .._multimethods import (
+    ndarray,
+    _identity_argreplacer,
+    _self_argreplacer,
+    _dtype_argreplacer,
+    mark_dtype,
+    _first_argreplacer,
+    _first2argreplacer,
+)
+
+
+@create_numpy(_first_argreplacer)
+@all_of_type(ndarray)
+def multi_dot(arrays):
+    return arrays
+
+
+def matrix_power_default(a, n):
+    if not ndim(a) == 2:
+        raise ValueError("Array must be two-dimensional")
+
+    if not isinstance(n, int):
+        raise ValueError("Exponent must be an integer.")
+
+    if a.shape[0] == a.shape[1]:
+        if n == 0:
+            return eye(a.shape[0], a.shape[1])
+        elif n < 0:
+            a = inv(a)
+
+        eigenvalues, eigenvectors = eig(a)
+
+        diagonal = diag(eigenvalues)
+
+        return multi_dot([eigenvectors, power(diagonal, abs(n)), inv(eigenvectors)])
+    else:
+        raise ValueError("The array must be square.")
+
+
+@create_numpy(_self_argreplacer, default=matrix_power_default)
+@all_of_type(ndarray)
+def matrix_power(a, n):
+    return (a,)
+
+
+@create_numpy(_self_argreplacer)
+@all_of_type(ndarray)
+def cholesky(a):
+    return (a,)
+
+
+@create_numpy(_self_argreplacer)
+@all_of_type(ndarray)
+def qr(a, mode="reduced"):
+    return (a,)
+
+
+@create_numpy(_self_argreplacer)
+@all_of_type(ndarray)
+def svd(a, full_matrices=True, compute_uv=True, hermitian=False):
+    return (a,)
+
+
+@create_numpy(_self_argreplacer)
+@all_of_type(ndarray)
+def eig(a):
+    return (a,)
+
+
+@create_numpy(_self_argreplacer)
+@all_of_type(ndarray)
+def eigh(a, UPLO="L"):
+    return (a,)
+
+
+@create_numpy(_self_argreplacer)
+@all_of_type(ndarray)
+def eigvals(a):
+    return (a,)
+
+
+@create_numpy(_self_argreplacer)
+@all_of_type(ndarray)
+def eigvalsh(a, UPLO="L"):
+    return (a,)
+
+
+def norm_default(x, ord=None, axis=None, keepdims=False):
+    x = abs(x)
+    if ndim(x) == 0:
+        norm = x
+    elif ndim(x) == 1:
+        if ord is None:
+            norm = sqrt(sum(square(x)))
+        elif ord == inf:
+            norm = max(x)
+        elif ord == NINF:
+            norm = min(x)
+        elif ord == 0:
+            norm = sum(x[x != 0])
```

Comment:
What does the `x[x != 0]` term do? If `x == 0` then it wouldn't contribute to the sum anyway.

Looking at the numpy docs, it says `sum(x != 0)` which would be the same as `count_nonzero(x)`.

Reply:
I though it was the sum of all non-zero elements which is what it does but you are right, it's a count and not a sum.